### PR TITLE
Make precompiled script plugin compilation compatible with the configuration cache

### PR DESCRIPTION
**Pull request by bamboo**

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test and merge

File: subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/precompiled/DefaultPrecompiledScriptPluginsSupport.kt [hunk 1]
```diff
@@ -343,23 +386,15 @@ fun Project.exposeScriptsAsGradlePlugins(scriptPlugins: List<PrecompiledScriptPl
...
-        }
-    }
+fun Project.collectScriptPluginFiles(): Set<File> =
+    gradlePlugin.pluginSourceSet.allSource.matching {
+        it.include("**/*.gradle.kts")
+    }.filter {
+        !it.isDirectory
+    }.also {
+        // Declare build configuration input
+        fileCollectionObserved(it, "Kotlin DSL")
```
Comment (eskatos) [inline subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/precompiled/DefaultPrecompiledScriptPluginsSupport.kt:396]: :question: This is tricky. Here we explicitly says that observing this FC is a build logic input. Shouldn't this be detected instead?

I'm not saying we need to do it now but it looks like a valid use case we may want to support eventually.

**Comment by bamboo** [comment]
@bot-gradle test and merge.

**Review by eskatos — APPROVED** [review]

**Comment by bamboo** [comment]
@bot-gradle test and merge

**Pull request merged**